File: languages/csharp/AppEncryption/AppEncryption.IntegrationTests/Regression/MetastoreMock.cs
using System;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.AppEncryption.IntegrationTests.TestHelpers;
using GoDaddy.Asherah.AppEncryption.Kms;
using GoDaddy.Asherah.AppEncryption.Persistence;
using GoDaddy.Asherah.Crypto.Engine.BouncyCastle;
using GoDaddy.Asherah.Crypto.Envelope;
using GoDaddy.Asherah.Crypto.Keys;
using Moq;
using Newtonsoft.Json.Linq;

namespace GoDaddy.Asherah.AppEncryption.IntegrationTests.Regression
{
    public static class MetastoreMock
    {
        private static readonly AeadEnvelopeCrypto Crypto = new BouncyAes256GcmCrypto();

        internal static Mock<IMetastore<JObject>> CreateMetastoreMock(
            Partition partition,
            KeyManagementService kms,
            KeyState metaIK,
            KeyState metaSK,
            CryptoKeyHolder cryptoKeyHolder,
            IMetastore<JObject> metastore)
        {
            CryptoKey systemKey = cryptoKeyHolder.SystemKey;

            Mock<IMetastore<JObject>> metastoreSpy = new Mock<IMetastore<JObject>>();

            metastoreSpy
                .Setup(x => x.Load(It.IsAny<string>(), It.IsAny<DateTimeOffset>()))
                .Returns<string, DateTimeOffset>(metastore.Load);
            metastoreSpy
                .Setup(x => x.LoadLatest(It.IsAny<string>()))
                .Returns<string>(metastore.LoadLatest);
            metastoreSpy
                .Setup(x => x.Store(It.IsAny<string>(), It.IsAny<DateTimeOffset>(), It.IsAny<JObject>()))
                .Returns<string, DateTimeOffset, JObject>(metastore.Store);

            if (metaSK != KeyState.Empty)
            {
                if (metaSK == KeyState.Retired)
                {
                    // We create a revoked copy of the same key
                    DateTimeOffset created = systemKey.GetCreated();
                    systemKey = systemKey
                        .WithKey(bytes => Crypto.GenerateKeyFromBytes(bytes, created, true));
                }

                EnvelopeKeyRecord systemKeyRecord = new EnvelopeKeyRecord(
                    systemKey.GetCreated(), null, kms.EncryptKey(systemKey), systemKey.IsRevoked());
                metastore.Store(
                    partition.SystemKeyId,
                    systemKeyRecord.Created,
                    systemKeyRecord.ToJson());
            }

            if (metaIK != KeyState.Empty)
            {
                CryptoKey intermediateKey = cryptoKeyHolder.IntermediateKey;
                if (metaIK == KeyState.Retired)
                {
                    // We create a revoked copy of the same key
                    DateTimeOffset created = intermediateKey.GetCreated();
                    intermediateKey = intermediateKey
                        .WithKey(bytes => Crypto.GenerateKeyFromBytes(bytes, created, true));
                }

                EnvelopeKeyRecord intermediateKeyRecord = new EnvelopeKeyRecord(
                    intermediateKey.GetCreated(),
                    new KeyMeta(partition.SystemKeyId, systemKey.GetCreated()),
                    Crypto.EncryptKey(intermediateKey, systemKey),
                    intermediateKey.IsRevoked());
                metastore.Store(
                    partition.IntermediateKeyId,
                    intermediateKeyRecord.Created,
                    intermediateKeyRecord.ToJson());
            }

            return metastoreSpy;
        }
    }
}

File: java/app-encryption/src/it/java/com/godaddy/asherah/regression/MetastoreMock.java
package com.godaddy.asherah.regression;

import com.godaddy.asherah.appencryption.Partition;
import com.godaddy.asherah.appencryption.envelope.EnvelopeKeyRecord;
import com.godaddy.asherah.appencryption.envelope.KeyMeta;
import com.godaddy.asherah.appencryption.kms.KeyManagementService;
import com.godaddy.asherah.appencryption.persistence.Metastore;
import com.godaddy.asherah.crypto.engine.bouncycastle.BouncyAes256GcmCrypto;
import com.godaddy.asherah.crypto.envelope.AeadEnvelopeCrypto;
import com.godaddy.asherah.crypto.keys.CryptoKey;
import com.godaddy.asherah.testhelpers.KeyState;
import org.json.JSONObject;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import java.time.Instant;

import static org.mockito.Mockito.*;

@SuppressWarnings("unchecked")
@ExtendWith(MockitoExtension.class)
final class MetastoreMock {
  private static final AeadEnvelopeCrypto CRYPTO = new BouncyAes256GcmCrypto();

  private MetastoreMock() { }

  static Metastore<JSONObject> createMetastoreMock(final Partition partition, final KeyManagementService kms,
      final Metastore<JSONObject> metastore, final KeyState metaIK, final KeyState metaSK,
      final CryptoKeyHolder cryptoKeyHolder) {

    Metastore<JSONObject> metastoreSpy = spy(metastore);
    CryptoKey systemKey = cryptoKeyHolder.getSystemKey();

    if (metaSK != KeyState.EMPTY) {
      if (metaSK == KeyState.RETIRED) {
        // We create a revoked copy of the same key
        Instant created = systemKey.getCreated();
        systemKey = systemKey.withKey((bytes) -> {
          return CRYPTO.generateKeyFromBytes(bytes, created, true);
        });
      }

      EnvelopeKeyRecord systemKeyRecord = new EnvelopeKeyRecord(systemKey.getCreated(),
        null, kms.encryptKey(systemKey), systemKey.isRevoked());
      metastoreSpy.store(partition.getSystemKeyId(), systemKeyRecord.getCreated(),
        systemKeyRecord.toJson());
    }

    if (metaIK != KeyState.EMPTY) {
      CryptoKey intermediateKey = cryptoKeyHolder.getIntermediateKey();
      if (metaIK == KeyState.RETIRED) {
        // We create a revoked copy of the same key
        Instant created = intermediateKey.getCreated();
        intermediateKey = intermediateKey.withKey((bytes) -> {
          return CRYPTO.generateKeyFromBytes(bytes, created, true);
        });
      }

      EnvelopeKeyRecord intermediateKeyRecord = new EnvelopeKeyRecord(intermediateKey.getCreated(),
        new KeyMeta(partition.getSystemKeyId(), systemKey.getCreated()),
        CRYPTO.encryptKey(intermediateKey, systemKey), intermediateKey.isRevoked());
      metastoreSpy.store(partition.getIntermediateKeyId(), intermediateKeyRecord.getCreated(),
        intermediateKeyRecord.toJson());
    }

    reset(metastoreSpy);
    return metastoreSpy;
  }
}

